File: 2025/day04.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.`[1:]

//go:embed day04.txt
var live string

type Point = tools.Point

type PointSet = map[Point]bool

var directions []Point = []Point{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}}

func makeset(grid []string) PointSet {
	rolls := make(PointSet)
	for y, row := range grid {
		for x, col := range row {
			if col == '@' {
				rolls[Point{x, y}] = true
			}
		}
	}
	return rolls
}

func part1(rolls PointSet) int {
	total := 0
	for pt := range rolls {
		count := 0
		for _, dpt := range directions {
			if rolls[pt.Add(dpt)] {
				count += 1
			}
		}
		if count < 4 {
			total++
		}
	}
	return total
}

func part2(rolls PointSet) int {
	removed := 0
	for true {
		nrolls := make(PointSet)
		for pt := range rolls {
			count := 0
			for _, dpt := range directions {
				if rolls[pt.Add(dpt)] {
					count += 1
				}
			}
			if count >= 4 {
				nrolls[pt] = true
			}
		}
		if len(rolls) == len(nrolls) {
			break
		}
		removed += len(rolls) - len(nrolls)
		rolls = nrolls
	}
	return removed
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := makeset(strings.Split(input, "\n"))

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}

File: 2025/day04.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cstdint>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
    "..@@.@@@@.\n"
    "@@@.@.@.@@\n"
    "@@@@@.@.@@\n"
    "@.@@@@..@.\n"
    "@@.@@@@.@@\n"
    ".@@@@@@@.@\n"
    ".@.@.@.@@@\n"
    "@.@@@.@@@@\n"
    ".@@@@@@@@.\n"
    "@.@.@@@.@."
);

#define DAY "day04"

bool DEBUG = false;
bool TEST = false;

typedef Point<int> point_t;
point_t directions[] = {
    point_t(-1,-1), point_t(-1,0), point_t(-1,1),
    point_t( 0,-1),                point_t( 0,1),
    point_t( 1,-1), point_t( 1,0), point_t( 1,1)
    };

typedef set<point_t> pointset_t;

int neighbors1( const pointset_t & rolls, const point_t & pt )
{
    int count = 0;
    for( auto & dxy : directions )
    {
        count += rolls.find(pt+dxy) != rolls.end();            
    }
    return count;
}

int neighbors( const pointset_t & rolls, const point_t & pt )
{
    return count_if(
        begin(directions), end(directions),
        [&rolls,pt](point_t & dxy) {
            return rolls.find(pt+dxy) != rolls.end();
        }
    );
}

int part1( const pointset_t & rolls )
{
    int total = 0;
    for( auto && pt : rolls )
        total += neighbors(rolls, pt) < 4;
    return total;
}

int part2 (pointset_t & rolls )
{
    int removed = 0;
	for( ;; )
    {
		pointset_t nrolls;
        copy_if(
            rolls.begin(),
            rolls.end(),
            inserter(nrolls, nrolls.begin()),
            [&rolls](const point_t & pt) { return neighbors(rolls, pt) >= 4; }
        );
        if( rolls.size() == nrolls.size() )
            break;
		removed += rolls.size() - nrolls.size();
		rolls = nrolls;
	}
	return removed;
}


int main( int argc, char ** argv )
{
    string name = *argv;
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");
    StringVector grid = split(input);
    pointset_t data;
    for( int y = 0; y < grid.size(); y++ )
        for( int x = 0; x < grid[y].size(); x++ )
            if( grid[y][x] == '@' )
                data.insert( point_t(x,y) );
    
    cout << "Part 1: " << part1(data) << "\n";
    cout << "Part 2: " << part2(data) << "\n";
}
